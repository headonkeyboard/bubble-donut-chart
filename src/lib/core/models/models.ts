export interface Point {
  x: number;
  y: number;
}

export interface RawData {
  id: string;
  label: string;
  weight: number;
  group: string;
}

export interface Section {
  id: string;
  ratio: number;
  bubbles: BubbleWithCoordsAndRadius[];
  startAngle: number;
  endAngle: number;
  gridLength: number;
  edges: GridEdges;
}

export interface GridEdges {
  top: Point[];
  right: Point[];
  bottom: Point[];
  left: Point[];
}

export type Coord = Point & { r: number };
export type BubbleWithCoordsAndRadius = RawData & Coord;
export type Grid = Array<number>;
