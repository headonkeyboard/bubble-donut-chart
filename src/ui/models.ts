export interface Section {
    bubbleCount: number;
    bubbleMinValue: number;
    bubbleMaxValue: number;
}
