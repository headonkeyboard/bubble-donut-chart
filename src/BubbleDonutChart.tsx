import React, {useEffect, useRef, FunctionComponent, useLayoutEffect} from 'react';
import * as d3 from 'd3';
import {Bubble, BubbleWithCoordsAndRadius} from "./lib/core/models/models";
import {BubbleDonut} from "./lib/core/bubble-donut";

type BubbleDonutChartProps = {
    bubbles: Bubble[]
}

function drawBubbles(selection: d3.Selection<SVGCircleElement, unknown, SVGSVGElement, unknown>, bubbles: Bubble[], width: number) {
    const diameter = width * 0.8;
    const radius = diameter / 2;
    const padding = (width - diameter);
    const offset = radius + padding / 2;

    const bubbleDonut = new BubbleDonut(50, radius, bubbles);

    const d3Bubbles: BubbleWithCoordsAndRadius[] = [];
    for (let [, bubbleSection] of bubbleDonut.sections) {
        bubbleSection.bubbles.forEach(bubble => {
            d3Bubbles.push(bubble);
        });
    }

    const groups = [...(new Set(d3.map(d3Bubbles, d => d.group)))];

    groups.sort((a, b) => {
       const sectionANumber = a.replace('group ', '');
       const sectionBNumber = b.replace('group ', '');

       return parseInt(sectionANumber, 10) - parseInt(sectionBNumber, 10);
    });

    var groupsColors = d3.scaleOrdinal().domain(groups)
        .range(d3.schemeSet2);

    const maxX: number = d3.max(d3Bubbles, (d) => {
        return radius + d.x + d.r;
    }) as number;

    var xScale = d3.scaleLinear()
        .domain([0, maxX])
        .range([0, width - padding]);

    var u: d3.Selection<SVGCircleElement, BubbleWithCoordsAndRadius, SVGSVGElement, unknown> = selection.data(d3Bubbles, d => (d as BubbleWithCoordsAndRadius).id);

    u.enter()
        .append('circle')
            .attr("opacity", 0)
            .attr("cx", xScale(radius))
            .attr("cy", xScale(radius))
        .merge(u)
            .transition()
            .duration(300)
            .attr("opacity", 1)
            .attr("cx", xScale(radius))
            .attr("cy", xScale(radius))
            .attr("cx", d => xScale((d as BubbleWithCoordsAndRadius).x + offset ))
            .attr("cy", d =>  xScale((d as BubbleWithCoordsAndRadius).y + offset ))
            .attr("fill", d => groupsColors(d.group) as string)
            .attr("r", d =>  xScale((d as BubbleWithCoordsAndRadius).r))

    u.exit()
        .transition()
        .duration(300)
        .attr("r", 0)
        .attr("opacity", 0)
        .remove()

}

const BubbleDonutChart: FunctionComponent<BubbleDonutChartProps> = ({ bubbles }) => {
    const svgRef = useRef<HTMLDivElement>(null);
    const wrapperWidth = useRef<number>(0);
    const d3SvgRef: React.MutableRefObject<d3.Selection<SVGSVGElement, unknown, null, undefined> | undefined> = useRef(undefined);

    useLayoutEffect(() => {
        if (null !== svgRef.current) {
            const elem = svgRef.current;

            wrapperWidth.current = Math.min(elem.clientWidth, elem.clientHeight);

            const svg = d3.select(svgRef.current)
                .append("svg")
                .attr("height", wrapperWidth.current)
                .attr("width", wrapperWidth.current)
            ;

            d3SvgRef.current = svg;

            return () => {
                svg.remove();
            }
        }
    }, []);

    useEffect(() => {
        if (d3SvgRef.current) {
            drawBubbles(d3SvgRef.current.selectAll("circle"), bubbles, wrapperWidth.current);
        }
    }, [bubbles]);

    useEffect(() => {
        function handleResize() {
            if (null !== svgRef.current && null != d3SvgRef.current) {
                const elem = svgRef.current;
                wrapperWidth.current = Math.min(elem.clientWidth, elem.clientHeight);

                d3SvgRef.current
                    .attr("height", wrapperWidth.current)
                    .attr("width", wrapperWidth.current)
                ;

                drawBubbles(d3SvgRef.current.selectAll("circle"), bubbles, wrapperWidth.current);
            }
        }

        window.addEventListener('resize', handleResize);

        return () => {
            window.removeEventListener('resize', handleResize);
        }
    }, [bubbles]);

    return (
        <div className="flex flex-1 justify-center items-center overflow-hidden" ref={svgRef}> {/* <div ref={svgRef} />*/}</div>

    );
};

export default BubbleDonutChart;
