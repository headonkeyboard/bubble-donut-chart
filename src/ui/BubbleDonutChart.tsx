import React, {
  useEffect,
  useRef,
  FunctionComponent,
  useLayoutEffect,
} from "react";
import * as d3 from "d3";
import {
  Bubble,
  BubbleWithCoordsAndRadius,
  Section,
} from "../lib/core/models/models";
import { BubbleDonut } from "../lib/core/bubble-donut";
import { groupColors } from "../utils/colors.utils";

type BubbleDonutChartProps = {
  bubbles: Bubble[];
};

/**
 * Returns a flat array of BubbleWithCoordsAndRadius extracted from each groups
 *
 * @param bubbleDonutSections
 */
const flattenDonutSectionBubbles = (
  bubbleDonutSections: Map<string, Section>
): BubbleWithCoordsAndRadius[] => {
  const d3Bubbles: BubbleWithCoordsAndRadius[] = [];

  for (let [, bubbleSection] of bubbleDonutSections) {
    bubbleSection.bubbles.forEach((bubble) => {
      d3Bubbles.push(bubble);
    });
  }

  return d3Bubbles;
};

function drawBubbles(
  selection: d3.Selection<SVGCircleElement, unknown, SVGSVGElement, unknown>,
  bubbles: Bubble[],
  width: number
) {
  const diameter = width * 0.8;
  const radius = diameter / 2;
  const padding = width - diameter;
  const offset = radius + padding / 2;

  const bubbleDonut = new BubbleDonut(radius * 0.2, radius, bubbles);
  const d3Bubbles = flattenDonutSectionBubbles(bubbleDonut.sections);

  const groups = [...new Set(d3.map(d3Bubbles, (d) => d.group))];
  var groupsColors = groupColors(groups.length);

  const maxX = d3.max(d3Bubbles, (d) => {
    return radius + d.x + d.r;
  }) as number;

  var xScale = d3
    .scaleLinear()
    .domain([0, maxX])
    .range([0, width - padding]);

  var u: d3.Selection<
    SVGCircleElement,
    BubbleWithCoordsAndRadius,
    SVGSVGElement,
    unknown
  > = selection.data(d3Bubbles, (d) => (d as BubbleWithCoordsAndRadius).id);

  u.enter()
    .append("circle")
    .attr("opacity", 0)
    .attr("cx", xScale(radius))
    .attr("cy", xScale(radius))
    .merge(u)
    .transition()
    .duration(300)
    .attr("opacity", 1)
    .attr("cx", xScale(radius))
    .attr("cy", xScale(radius))
    .attr("cx", (d) => xScale((d as BubbleWithCoordsAndRadius).x + offset))
    .attr("cy", (d) => xScale((d as BubbleWithCoordsAndRadius).y + offset))
    .attr("fill", (d) => groupsColors(d.group) as string)
    .attr("r", (d) => xScale((d as BubbleWithCoordsAndRadius).r));

  u.exit().transition().duration(300).attr("r", 0).attr("opacity", 0).remove();
}

const BubbleDonutChart: FunctionComponent<BubbleDonutChartProps> = ({
  bubbles,
}) => {
  const svgRef = useRef<HTMLDivElement>(null);
  const wrapperWidth = useRef<number>(0);
  const d3SvgRef: React.MutableRefObject<
    d3.Selection<SVGSVGElement, unknown, null, undefined> | undefined
  > = useRef(undefined);

  useLayoutEffect(() => {
    if (null !== svgRef.current) {
      const elem = svgRef.current;

      wrapperWidth.current = Math.min(elem.clientWidth, elem.clientHeight);

      const svg = d3
        .select(svgRef.current)
        .append("svg")
        .attr("height", wrapperWidth.current)
        .attr("width", wrapperWidth.current);
      d3SvgRef.current = svg;

      return () => {
        svg.remove();
      };
    }
  }, []);

  useEffect(() => {
    if (d3SvgRef.current) {
      drawBubbles(
        d3SvgRef.current.selectAll("circle"),
        bubbles,
        wrapperWidth.current
      );
    }
  }, [bubbles]);

  useEffect(() => {
    function handleResize() {
      if (null !== svgRef.current && null != d3SvgRef.current) {
        const elem = svgRef.current;
        wrapperWidth.current = Math.min(elem.clientWidth, elem.clientHeight);

        d3SvgRef.current
          .attr("height", wrapperWidth.current)
          .attr("width", wrapperWidth.current);

        drawBubbles(
          d3SvgRef.current.selectAll("circle"),
          bubbles,
          wrapperWidth.current
        );
      }
    }

    window.addEventListener("resize", handleResize);

    return () => {
      window.removeEventListener("resize", handleResize);
    };
  }, [bubbles]);

  return (
    <div
      className="flex flex-1 justify-center items-center overflow-hidden"
      ref={svgRef}
    ></div>
  );
};

export default BubbleDonutChart;
