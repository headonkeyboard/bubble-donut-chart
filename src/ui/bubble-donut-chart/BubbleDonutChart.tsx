import React, {
  useEffect,
  useRef,
  FunctionComponent,
  useLayoutEffect,
} from "react";
import { RawData } from "../../lib/core/models";
import {
  createD3SvgRoot,
  D3SVGSelection,
  layoutBubbles,
  resizeD3ViewportAndBubbles,
} from "./utils/bubble-donut-d3-layout.utils";

type BubbleDonutChartProps = {
  rawData: RawData[];
};

const BubbleDonutChart: FunctionComponent<BubbleDonutChartProps> = ({
  rawData,
}) => {
  const svgWrapperRef = useRef<HTMLDivElement>(null);
  const d3BubbleSelectionRef: React.MutableRefObject<
    D3SVGSelection | undefined
  > = useRef(undefined);

  useLayoutEffect(() => {
    if (null !== svgWrapperRef.current) {
      const size = Math.min(
        svgWrapperRef.current.clientWidth,
        svgWrapperRef.current.clientHeight
      );
      d3BubbleSelectionRef.current = createD3SvgRoot(
        svgWrapperRef.current,
        size
      );

      return () => {
        if (d3BubbleSelectionRef.current) {
          d3BubbleSelectionRef.current.remove();
        }
      };
    }
  }, []);

  /* Bubbles need to layout on every changes */
  useEffect(() => {
    if (svgWrapperRef.current) {
      const size = Math.min(
        svgWrapperRef.current.clientWidth,
        svgWrapperRef.current.clientHeight
      );

      // priorize ui interactions as bubble layout might take some time
      setTimeout(() => {
        if (d3BubbleSelectionRef.current) {
          layoutBubbles(d3BubbleSelectionRef.current, size, rawData);
        }
      });
    }
  }, [rawData]);

  useEffect(() => {
    function handleResize() {
      if (
        null !== svgWrapperRef.current &&
        null != d3BubbleSelectionRef.current
      ) {
        const newSize = Math.min(
          svgWrapperRef.current.clientWidth,
          svgWrapperRef.current.clientHeight
        );

        resizeD3ViewportAndBubbles(d3BubbleSelectionRef.current, newSize);
      }
    }

    window.addEventListener("resize", handleResize);

    return () => {
      window.removeEventListener("resize", handleResize);
    };
  }, [rawData]);

  return (
    <div
      className="flex flex-1 justify-center items-center overflow-hidden"
      ref={svgWrapperRef}
    />
  );
};

export default BubbleDonutChart;
