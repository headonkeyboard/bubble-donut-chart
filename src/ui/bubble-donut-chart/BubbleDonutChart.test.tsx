import React from "react";
import {render} from "@testing-library/react";
import BubbleDonutChart from "./BubbleDonutChart";
import {RawData} from "../../lib/core/models/models";

const data: RawData[] = [
    { id: '1', group: 'group 1', label: 'test', weight: 10 },
    { id: '2', group: 'group 1', label: 'test', weight: 10 },
    { id: '3', group: 'group 2', label: 'test', weight: 10 },
    { id: '4', group: 'group 2', label: 'test', weight: 10 },
];

describe('BubbleDonutChart', () => {
    test('BubbleDonutChart should render circles', (done) => {
        const {container} = render(
            <BubbleDonutChart rawData={data} />
        );

        // wait for animations
        setTimeout(() => {
            expect(container).toMatchSnapshot();
            done();
        }, 1000);
    });
});
