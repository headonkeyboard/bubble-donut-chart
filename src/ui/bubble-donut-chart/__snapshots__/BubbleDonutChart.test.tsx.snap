// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BubbleDonutChart BubbleDonutChart should render circles 1`] = `
<div>
  <div
    class="flex flex-1 justify-center items-center overflow-hidden"
  >
    <svg
      height="0"
      width="0"
    >
      <circle
        cx="0"
        cy="0"
        fill="#66c2a5"
        opacity="1"
        r="0"
      />
      <circle
        cx="0"
        cy="0"
        fill="#66c2a5"
        opacity="1"
        r="0"
      />
      <circle
        cx="0"
        cy="0"
        fill="#fc8d62"
        opacity="1"
        r="0"
      />
      <circle
        cx="0"
        cy="0"
        fill="#fc8d62"
        opacity="1"
        r="0"
      />
    </svg>
  </div>
</div>
`;
